README.md
# fl_mlkit_identify_language

## Identify the language of text with ML Kit on Android And IOS

* 如果调用成功，则会返回[BCP-47语言代码](https://en.wikipedia.org/wiki/IETF_language_tag)
  tag和confidence,如果没有检测到语言，则返回通过代码[und](https://www.rfc-editor.org/info/bcp47#section-4.1)(undetermined)。

* If the call succeeds, a [BCP-47 language code](https://en.wikipedia.org/wiki/IETF_language_tag) is
  passed to the success listener, indicating the language of the text. If no language is confidently
  detected, the code [und](https://www.rfc-editor.org/info/bcp47#section-4.1)(undetermined) is
  passed.

* [supported languages](https://developers.google.com/ml-kit/language/identification/langid-support)

```dart

void func() {
  FlMlKitIdentifyLanguage mlKitIdentifyLanguage = FlMlKitIdentifyLanguage();

  /// Identify language
  /// If the call succeeds, a BCP-47 language code is passed to the success listener,
  /// indicating the language of the text. If no language is confidently detected,
  /// the code und (undetermined) is passed.
  mlKitIdentifyLanguage.identifyLanguage('text');

  /// Identify possible languages
  mlKitIdentifyLanguage.identifyPossibleLanguages('text');

  /// Set confidence
  /// Confidence must >=0.01 and must <=1
  /// Before you set confidence, each identifyLanguage will use the previous confidence or the default 0.5
  mlKitIdentifyLanguage.setConfidence(0.75);

  /// Be sure to call this method when you are no longer using a collapsible
  mlKitIdentifyLanguage.dispose();
}

```

<img src="https://github.com/Wayaer/fl_mlkit_identify_language/raw/main/res/example.png" width="60%"/>

CHANGELOG.md
## 0.0.2
* Initial release.


example/lib/main.dart
import 'package:fl_mlkit_identify_language/fl_mlkit_identify_language.dart';
import 'package:flutter/material.dart';

import 'package:flutter_waya/flutter_waya.dart';

void main() {
  runApp(ExtendedWidgetsApp(home: const _App()));
}

class _App extends StatefulWidget {
  const _App({Key? key}) : super(key: key);

  @override
  _AppState createState() => _AppState();
}

class _AppState extends State<_App> {
  ValueNotifier<List<IdentifiedLanguageModel>> identifiedLanguageModel =
      ValueNotifier<List<IdentifiedLanguageModel>>([]);
  TextEditingController controller = TextEditingController();
  FlMlKitIdentifyLanguage mlKitIdentifyLanguage = FlMlKitIdentifyLanguage();

  @override
  Widget build(BuildContext context) {
    return ExtendedScaffold(
        isScroll: true,
        appBar: AppBarText('Fl MlKit Identify Language'),
        mainAxisAlignment: MainAxisAlignment.center,
        padding: const EdgeInsets.all(30),
        children: [
          TextField(
              controller: controller,
              maxLength: 500,
              maxLines: 4,
              onSubmitted: (value) {
                context.focusNode();
              },
              decoration: const InputDecoration(
                  border: OutlineInputBorder(), hintText: 'Please enter text')),
          ExtendedFutureBuilder<double>(
              future: () async => mlKitIdentifyLanguage.confidence,
              onWaiting: (_) => const CircularProgressIndicator(),
              onDone: (_, double value, reset) {
                return ElevatedText(
                    onPressed: () async {
                      context.focusNode();
                      final confidence = await selectConfidence();
                      if (confidence != null) {
                        final state = await mlKitIdentifyLanguage
                            .setConfidence(confidence);
                        if (state) reset();
                      }
                    },
                    text: 'Click Modify Confidence : ' + value.toString());
              }),
          ElevatedText(text: 'Identify Language', onPressed: identifyLanguage),
          ElevatedText(
              text: 'Get Native Confidence',
              onPressed: () {
                mlKitIdentifyLanguage.getCurrentConfidence().then((value) {
                  showToast(value.toString());
                });
              }),
          ElevatedText(
              text: 'Identify Possible Language',
              onPressed: identifyPossibleLanguages),
          const SizedBox(height: 20),
          ValueListenableBuilder(
              valueListenable: identifiedLanguageModel,
              builder: (_, List<IdentifiedLanguageModel> value, __) {
                return Column(
                    children: value.builder((item) => Container(
                        margin: const EdgeInsets.symmetric(vertical: 10),
                        width: double.infinity,
                        padding: const EdgeInsets.all(10),
                        decoration: BoxDecoration(
                            color: Colors.white,
                            borderRadius: BorderRadius.circular(6)),
                        child: RText(
                            texts: [
                              'confidence：',
                              item.confidence.toString(),
                              '      languageTag：',
                              item.languageTag,
                            ],
                            textAlign: TextAlign.start,
                            styles: const [
                              BTextStyle(
                                  color: Colors.black87,
                                  fontWeight: FontWeight.w500),
                              BTextStyle(
                                  color: Colors.black,
                                  fontWeight: FontWeight.bold),
                              BTextStyle(
                                  color: Colors.black87,
                                  fontWeight: FontWeight.w500),
                              BTextStyle(
                                  color: Colors.black,
                                  fontWeight: FontWeight.bold),
                            ]))));
              })
        ]);
  }

  Future<void> identifyLanguage() async {
    if (controller.text.isEmpty) {
      showToast('Please enter the text');
      return;
    }
    context.focusNode();
    final data = await mlKitIdentifyLanguage.identifyLanguage(controller.text);
    if (data != null) {
      identifiedLanguageModel.value = [data];
      setState(() {});
    }
  }

  Future<void> identifyPossibleLanguages() async {
    if (controller.text.isEmpty) {
      showToast('Please enter the text');
      return;
    }
    context.focusNode();
    final data =
        await mlKitIdentifyLanguage.identifyPossibleLanguages(controller.text);
    identifiedLanguageModel.value = data;
    setState(() {});
  }

  Future<double?> selectConfidence() {
    var confidences = [0.01, 0.1, 0.25, 0.5, 0.75, 1.0];
    return showBottomPopup<double?>(
        options: BottomSheetOptions(isScrollControlled: false),
        widget: Universal(
          mainAxisSize: MainAxisSize.min,
          children: confidences.builder((item) => Universal(
              width: double.infinity,
              decoration: const BoxDecoration(
                  border: Border(
                      bottom: BorderSide(color: Colors.black12, width: 0.5))),
              onTap: () {
                pop(item);
              },
              padding: const EdgeInsets.symmetric(vertical: 8),
              child: Text(item.toString(), textAlign: TextAlign.center))),
          padding: const EdgeInsets.all(20),
          decoration: const BoxDecoration(
              color: Colors.white,
              borderRadius: BorderRadius.horizontal(
                  left: Radius.circular(10), right: Radius.circular(10))),
        ));
  }
}

class AppBarText extends AppBar {
  AppBarText(String text, {Key? key})
      : super(
            key: key,
            elevation: 0,
            title: BText(text,
                color: Colors.white, fontSize: 18, fontWeight: FontWeight.bold),
            centerTitle: true);
}

class ElevatedText extends StatelessWidget {
  const ElevatedText({Key? key, this.onPressed, required this.text})
      : super(key: key);
  final VoidCallback? onPressed;
  final String text;

  @override
  Widget build(BuildContext context) =>
      ElevatedButton(onPressed: onPressed, child: Text(text));
}

class ElevatedIcon extends StatelessWidget {
  const ElevatedIcon({Key? key, this.onPressed, required this.icon})
      : super(key: key);
  final VoidCallback? onPressed;
  final IconData icon;

  @override
  Widget build(BuildContext context) => ElevatedButton(
      onPressed: onPressed, child: Icon(icon, color: Colors.white));
}


android/src/main/kotlin/fl/mlkit/identify/language/FlMlKitIdentifyLanguagePlugin.kt
package fl.mlkit.identify.language

import androidx.annotation.NonNull
import com.google.mlkit.nl.languageid.IdentifiedLanguage
import com.google.mlkit.nl.languageid.LanguageIdentification
import com.google.mlkit.nl.languageid.LanguageIdentificationOptions
import com.google.mlkit.nl.languageid.LanguageIdentifier

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlMlKitIdentifyLanguagePlugin */
class FlMlKitIdentifyLanguagePlugin : FlutterPlugin, MethodCallHandler {

    private lateinit var channel: MethodChannel

    private var languageIdentifier: LanguageIdentifier? = null
    private var options: LanguageIdentificationOptions? = null
    private var currentConfidence: Float = 0.5f

    override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
        channel = MethodChannel(flutterPluginBinding.binaryMessenger, "fl_mlkit_identify_language")
        channel.setMethodCallHandler(this)
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
        when (call.method) {
            "identifyLanguage" -> {
                val text = call.arguments as String
                getLanguageIdentification().identifyLanguage(text)
                    .addOnSuccessListener { languageCode ->
                        result.success(languageCode)
                    }.addOnFailureListener {
                        result.success(null)
                    }
            }
            "identifyPossibleLanguages" -> {
                val text = call.arguments as String
                getLanguageIdentification().identifyPossibleLanguages(text)
                    .addOnSuccessListener { identifiedLanguages ->
                        result.success(identifiedLanguages.map { model -> model.data })
                    }
                    .addOnFailureListener {
                        result.success(null)
                    }
            }
            "setConfidence" -> {
                currentConfidence = (call.arguments as Double).toFloat()
                dispose()
                result.success(true)
            }
            "getCurrentConfidence" -> result.success(currentConfidence)
            "dispose" -> {
                dispose()
                result.success(true)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private val IdentifiedLanguage.data: Map<String, Any?>
        get() = mapOf(
            "languageTag" to languageTag,
            "confidence" to confidence,
        )

    private fun dispose() {
        options = null
        languageIdentifier?.close()
        languageIdentifier = null
    }


    private fun getLanguageIdentification(): LanguageIdentifier {
        if (languageIdentifier == null) {
            languageIdentifier = LanguageIdentification.getClient(getOptions())
        }
        return languageIdentifier!!
    }

    private fun getOptions(): LanguageIdentificationOptions {
        if (options == null) {
            options = LanguageIdentificationOptions.Builder()
                .setConfidenceThreshold(currentConfidence)
                .build()
        }
        return options!!
    }


    override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
    }
}


android/src/main/kotlin/fl/mlkit/identify/language/FlMlkitIdentifyLanguagePlugin.kt
package fl.mlkit.identify.language

import androidx.annotation.NonNull

import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import io.flutter.plugin.common.MethodChannel.Result

/** FlMlkitIdentifyLanguagePlugin */
class FlMlkitIdentifyLanguagePlugin: FlutterPlugin, MethodCallHandler {
  /// The MethodChannel that will the communication between Flutter and native Android
  ///
  /// This local reference serves to register the plugin with the Flutter Engine and unregister it
  /// when the Flutter Engine is detached from the Activity
  private lateinit var channel : MethodChannel

  override fun onAttachedToEngine(@NonNull flutterPluginBinding: FlutterPlugin.FlutterPluginBinding) {
    channel = MethodChannel(flutterPluginBinding.binaryMessenger, "fl_mlkit_identify_language")
    channel.setMethodCallHandler(this)
  }

  override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: Result) {
    if (call.method == "getPlatformVersion") {
      result.success("Android ${android.os.Build.VERSION.RELEASE}")
    } else {
      result.notImplemented()
    }
  }

  override fun onDetachedFromEngine(@NonNull binding: FlutterPlugin.FlutterPluginBinding) {
    channel.setMethodCallHandler(null)
  }
}


lib/fl_mlkit_identify_language.dart
import 'dart:async';

import 'package:flutter/services.dart';

const MethodChannel _channel = MethodChannel('fl_mlkit_identify_language');

class FlMlKitIdentifyLanguage {
  factory FlMlKitIdentifyLanguage() =>
      _singleton ??= FlMlKitIdentifyLanguage._();

  FlMlKitIdentifyLanguage._();

  static FlMlKitIdentifyLanguage? _singleton;

  double _confidence = 0.5;

  double get confidence => _confidence;

  /// Identify language
  Future<IdentifiedLanguageModel?> identifyLanguage(String text) async {
    final String? tag =
        await _channel.invokeMethod<String?>('identifyLanguage', text);
    if (tag != null) {
      return IdentifiedLanguageModel(confidence: _confidence, languageTag: tag);
    }
    return null;
  }

  /// Identify possible languages
  Future<List<IdentifiedLanguageModel>> identifyPossibleLanguages(
      String text) async {
    final List<Map<dynamic, dynamic>>? list =
        await _channel.invokeListMethod<Map<dynamic, dynamic>>(
            'identifyPossibleLanguages', text);
    if (list != null && list.isNotEmpty) {
      return list.map((data) => IdentifiedLanguageModel.fromMap(data)).toList();
    }
    return [];
  }

  /// Set confidence
  Future<bool> setConfidence(double confidence) async {
    assert(confidence >= 0.01 && confidence <= 1);
    final bool? state =
        await _channel.invokeMethod<bool?>('setConfidence', confidence);
    if (state == true) _confidence = confidence;
    return state ?? false;
  }

  /// Get naitve confidence
  Future<double> getCurrentConfidence() async {
    final double? confidence =
        await _channel.invokeMethod<double?>('getCurrentConfidence');
    if (confidence != null) _confidence = confidence;
    return confidence ?? _confidence;
  }

  /// Be sure to call this method when you are no longer using a collapsible
  Future<bool> dispose(double confidence) async =>
      (await _channel.invokeMethod<bool?>('dispose')) ?? false;
}

class IdentifiedLanguageModel {
  IdentifiedLanguageModel(
      {required this.languageTag, required this.confidence});

  IdentifiedLanguageModel.fromMap(Map<dynamic, dynamic> data)
      : languageTag = data['languageTag'] as String,
        confidence = data['confidence'] as double;

  /// If the call succeeds, a BCP-47 language code(https://en.wikipedia.org/wiki/IETF_language_tag) is passed to the success listener,
  /// indicating the language of the text. If no language is confidently detected,
  /// the code und (undetermined) is passed.
  String languageTag;

  /// Identified confidence
  double confidence;
}
